### ISSUE TITLE
question about most significant modified bases

### ISSUE THREAD

**Issue by ma-diroma** [nanoporetech/tombo#37]
Hi,

could you please tell me how the most significant modified bases are selected for plotting with `plot_most_significant`? I thought it could take into consideration the highest fractions of modified bases (reported in the wiggle file), but I see it is not possible.
Moreover, after an RNA analysis with tombo, transcriptomic coordinates are plotted. I tried to go back to the genomic coordinates, but there is no correspondence between the sequences, e.g.

ENST00000012049.9_1 position 1776 corresponds to chr19:46197516. I show you what I see in my plot and in UCSC genomic view of this region

<img width="1328" alt="schermata 2018-03-02 alle 13 33 02" src="https://user-images.githubusercontent.com/13780606/36899187-49bed18c-1e1e-11e8-9ea4-ebf6904c8230.png">

<img width="1426" alt="schermata 2018-03-02 alle 13 33 46" src="https://user-images.githubusercontent.com/13780606/36899213-6e2cfe22-1e1e-11e8-87a2-6396985a6490.png">

Could you please help me to understand?
Thanks,
Maria Angela

**Comment by marcus1487** [comment]
For the first point, the plot_most_significant command takes a Tombo stats file, not a wiggle file. The wiggle files are intended only as a final output from Tombo. This might be your issue, but let me know if that doesn't fix your plotting issue.

For the second point, usually this has to do with discordant genome/transcriptome versions. I know these issues can be quite complex. I am working on a pipeline that will allow conversion between transcriptome and genome coordinates and will post this on the documentation when it is completed ([here](https://nanoporetech.github.io/tombo/rna.html)). In order to make sure this is indeed not a Tombo bug, could you check that the transcriptome coordinates match (potentially via a browser that allows user fasta files; e.g. IGV). Another potential issue is the coordinate conversion. There may be a splice junction making the coordinate conversion more complex. You could test this by blasting a longer bit of sequence from a Tombo plot to your genome of interest. Please do let me know if you find discordant positions with the actual transcriptome fasta file provided for Tombo processing.

**Comment by ma-diroma** [comment]
Yes, I know that `plot_most_significant` command takes a Tombo stats file, but I expected to find a correlation with the fraction.wiggle output file (generated from the same Tombo stats file), that is the sites with the highest fraction of modified bases should be plotted as most significant regions, but clearly I am wrong. So, I would like to understand how `plot_most_significant` selects the most significant modified bases.
For the second point, I can confirm it is not a question of discordant versions
<img width="982" alt="schermata 2018-03-02 alle 16 59 09" src="https://user-images.githubusercontent.com/13780606/36908318-2d7e41d4-1e3b-11e8-84ed-c4bc81b52f8e.png">
that is exactly the reference transcriptome I used and the transcriptome coordinates are converted automatically by UCSC. However the sequence shown above is clearly different from that plotted.
This is longer bit if sequence around that modified base
<img width="1348" alt="schermata 2018-03-02 alle 17 36 36" src="https://user-images.githubusercontent.com/13780606/36910217-672476a6-1e40-11e8-84b6-4fd90d021c88.png">
no blast/blat result.

**Comment by marcus1487** [comment]
I think you are indeed interpreting this correctly. The central position plotted from the `plot_most_significant` command is the positions with the largest identified fraction of modified bases from the tombo stats file. Then subsequent pages of the PDF go through the next most modified sites from the stats file. I think I may not be understanding your issue exactly. Would you mind explaining what you mean by find a correlation with the wiggle file? Do the plotted regions not appear to be modified? Note here that the default for de novo testing (comparison to the canonical model) is to combine the test statistics across a moving window of 3 bases (via the `--fishers-method-context` option). Thus a read might be not be quite significant at 3 consecutive positions, but then when the p-values are combined they become significant, and this can be hard to see visually. You can read more about this in the [documentation here](https://nanoporetech.github.io/tombo/modified_base_detection.html#canonical-sample-comparison-method). It can be more visually clear if you test for significant shifts with the `--fishers-method-context` set to `0`, but I have found that the accuracy is better with a moving window of 3 as signal is generally shifted at several locations around a modified base.

For the mismatching sequence, can you confirm that the sequence does not match that sequence provided to Tombo (via IGV or sequence extraction from the transcriptome reference FASTA).

I can confirm that this larger bit of sequence is part of the referenced gene in ENSMBL:
![image](https://user-images.githubusercontent.com/13755444/36911103-69602fbc-1df7-11e8-9256-54a707c1e77f.png)

It also appears as though this bit of sequence does not match the associated genomic sequence, so it is likely that this stretch spans a splice junction further complicating the coordinate identification.

Please let me know if you have additional questions concerning your analysis.

**Comment by ma-diroma** [comment]
I expected to find plotted those positions with a fraction value=1.0, while I found also very low values (in this case 0.027).
About coordinates within plot, are they 0-based?

**Comment by marcus1487** [comment]
Ah there is certainly something wrong if you are getting `1.0` values in the fraction wiggle output files, but 0.027 listed in the title of the `plot_most_significant` command results.

For the genome-anchored plots, they are all 1-based. I will add a note to add this to the documentation.

**Comment by ma-diroma** [comment]
I am sorry for bothering you. What do you mean for something wrong? Why 1.0 values are not expected?

**Comment by marcus1487** [comment]
I mean that there is a bug in Tombo if you are getting fraction values of `1.0` in you wiggle files, but those values are not represented in your `plot_most_significant` results. This was my interpretation of your previous comment. Is this the case?

If this is not the case, then why are you expecting fraction values of `1.0`?

**Comment by ma-diroma** [comment]
(Please, forget the previous case)
I have more than 50,000 `1.0` values in my fraction.wiggle.
The first region plotted by `plot_most_significant` is this

<img width="1330" alt="schermata 2018-03-02 alle 23 49 35" src="https://user-images.githubusercontent.com/13780606/36925719-71c91f2e-1e74-11e8-8562-9d82ebe42fdc.png">

and the corresponding values within fraction.wiggle are these

```
grep ENST00000370474.3_1 myfile.fraction_modified_reads.plus.wig -A 20
variableStep chrom=ENST00000370474.3_1 span=1
4 1.0
5 0.0
6 0.0
7 0.5
8 0.3333
9 0.3333
10 0.3333
11 0.3333
12 0.0
13 0.0
14 0.2
15 0.1667
16 0.1667
17 0.1667
18 0.1667
19 0.3333
20 0.3333
21 0.4286
22 0.1429
23 0.0
```

I have plotted 30 most significant regions and I think there is a bias towards the starting of each transcript. In all cases the first 20-25 bases are shown.

**Comment by marcus1487** [comment]
This is a problem I had started to address in this release, but the next release will incorporate this fix over all applications of Tombo statistics files. The problem boils down to how to deal with results over variable coverage. I have moved away from reporting combined p-values at each genomic location since very high coverage will usually have a number of outliers that will skew the results  in favor of high coverage positions (which have some outliers on average). Thus I switched to using the fraction of reads passing some fixed threshold at each genomic position. This has the opposite effect, where the fraction of modified reads at a genomic position is highly variable for low coverage regions.

I have now added a `--coverage-dampen-counts` option to the `write_wiggles` command in this release. This has the effect of adding psudo-counts (for modified and unmodified reads) to each genomic position before adding in the read observations. Thus very low coverage regions show a dampened fraction of modified bases, while high coverage regions are not effected much by the small pseudo-counts. I hope to have this option applied to all commands which accept the Tombo statistics file in the next release. I will also (most likely) have this value set as the default in the next release so that interpreting these fraction values is easier over low coverage regions by default. I hope this makes sense!
